In [11]:
#Import relevant libraries and define the Data path
import pandas as pd
import matplotlib.pyplot as plt
import glob

DATA_PATH = '../data/DayAheadPrices_12.1.D/*.csv'


files = glob.glob(DATA_PATH)

no2_dataframes = []
for f in files:
    csv = pd.read_csv(f, delimiter = '\t') 
    csv = csv[csv['MapCode'] == 'NO2']
    csv = csv[['DateTime', 'Price']]
    csv.rename(columns = {'DateTime' : 'Date', 'Price' : 'Price'}, inplace = True)
    no2_dataframes.append(csv)

no2_data = pd.concat(no2_dataframes, ignore_index = True)
no2_data['Date'] = pd.to_datetime(no2_data['Date'])



In [12]:
#Doing the same procedure in last task for collecting relevant German data
ger_dataframes = []
for f in files:
    csv = pd.read_csv(f, delimiter = '\t') 
    csv = csv[csv['MapCode'] == 'DE_LU']
    csv = csv[['DateTime', 'Price']]
    csv.rename(columns = {'DateTime' : 'Date', 'Price' : 'Price'}, inplace = True)
    
    ger_dataframes.append(csv)


ger_data = pd.concat(ger_dataframes, ignore_index = True)
ger_data['Date'] = pd.to_datetime(ger_data['Date'])

In [14]:
# Extract year from DateTime in both created dataframes and add it as a new column
no2_data['Year'] = no2_data['Date'].dt.year
ger_data['Year'] = ger_data['Date'].dt.year

In [22]:
#Initialize an empty list to store our results, enabling us to create a new dataframe later...

results = []
# NB! Keeping all calculations and code inside the for loop to get stats for each year
for year in range(2019, 2024):
    # Filter relevant data for the specific years
    yearly_no2 = no2_data[no2_data['Year'] == year]['Price']
    yearly_ger = ger_data[ger_data['Year'] == year]['Price']

#Creating two dictionaries to store the calculated statistics for the specific years in NO2 and Germany, rounding to two decimals
    stats_no2 = {
        'Year' : year,
        'Country' : 'NO2', 
        'Mean' : round(yearly_no2.mean(), 2),
        'Median' : round(yearly_no2.median(), 2),
        'Std Dev' : round(yearly_no2.std(), 2),
        'Min' : round(yearly_no2.min(), 2), 
        'Max' : round(yearly_no2.max(), 2)
}

    stats_ger = {
        'Year' : year,
        'Country' : 'Germany', 
        'Mean' : round(yearly_ger.mean(), 2),
        'Median' : round(yearly_ger.median(), 2),
        'Std Dev' : round(yearly_ger.std(), 2),
        'Min' : round(yearly_ger.min(), 2), 
        'Max' : round(yearly_ger.max(), 2)    
    }

    #Appending all results from dictionaries to the initialized list
    results.append(stats_no2)
    results.append(stats_ger)
results

[{'Year': 2019,
  'Country': 'NO2',
  'Mean': 39.26,
  'Median': 38.81,
  'Std Dev': 8.23,
  'Min': 5.86,
  'Max': 109.45},
 {'Year': 2019,
  'Country': 'Germany',
  'Mean': 37.67,
  'Median': 38.06,
  'Std Dev': 15.52,
  'Min': -90.01,
  'Max': 121.46},
 {'Year': 2020,
  'Country': 'NO2',
  'Mean': 9.29,
  'Median': 6.95,
  'Std Dev': 8.26,
  'Min': -1.73,
  'Max': 99.92},
 {'Year': 2020,
  'Country': 'Germany',
  'Mean': 30.47,
  'Median': 30.99,
  'Std Dev': 17.5,
  'Min': -83.94,
  'Max': 200.04},
 {'Year': 2021,
  'Country': 'NO2',
  'Mean': 75.11,
  'Median': 59.34,
  'Std Dev': 47.22,
  'Min': -1.97,
  'Max': 600.16},
 {'Year': 2021,
  'Country': 'Germany',
  'Mean': 130.41,
  'Median': 90.44,
  'Std Dev': 104.8,
  'Min': -149.99,
  'Max': 809.5},
 {'Year': 2022,
  'Country': 'NO2',
  'Mean': 211.28,
  'Median': 181.13,
  'Std Dev': 125.85,
  'Min': 0.04,
  'Max': 844.0},
 {'Year': 2022,
  'Country': 'Germany',
  'Mean': 237.1,
  'Median': 208.53,
  'Std Dev': 148.24,
  'Min': -

In [23]:
#Creating a dataframe for the statistics we calculated above
stats_df = pd.DataFrame(results)
stats_df

Year  Country    Mean  Median  Std Dev     Min      Max
0  2019      NO2   39.26   38.81     8.23    5.86   109.45
1  2019  Germany   37.67   38.06    15.52  -90.01   121.46
2  2020      NO2    9.29    6.95     8.26   -1.73    99.92
3  2020  Germany   30.47   30.99    17.50  -83.94   200.04
4  2021      NO2   75.11   59.34    47.22   -1.97   600.16
5  2021  Germany  130.41   90.44   104.80 -149.99   809.50
6  2022      NO2  211.28  181.13   125.85    0.04   844.00
7  2022  Germany  237.10  208.53   148.24 -149.99  2999.99
8  2023      NO2   79.44   80.45    36.28  -61.84   261.85
9  2023  Germany   95.07   95.90    59.16 -500.00   800.00

In [24]:
#Export the dataframe into an excel file
stats_df.to_excel('table_task2.xlsx')

Conclusion: 
By looking into the columns for the years after NordLink was deployed (2021,2022), we see that the electricity prices overall has increased rapidly. The 'std dev' column and the widening of 'min' and 'max' reflects this.
The average prices in both NO2 and Germany have generally trended upwards. For instance, the mean price for NO2 increases from 39.26 in 2019 to a peak of 211.28 in 2022, while Germany’s mean price rises from 37.67 in 2019 to 237.10 in 2022. This shows a extreme growing trend in both regions. 

Germany’s volatility (standard deviation) and extremes (min and max values) are slightly higher than in NO2, especially in 2022, where the maximum price reaches nearly 3000 and the minimum is a large negative value (-149.99). This might be due to Germany’s shift from more stable energy sources like oil, gas, and nuclear power to less consistent renewable sources like wind and solar.
